### PR TITLE
[Snyk] Security upgrade transformers from 4.51.3 to 5.3.0

### PR DESCRIPTION
**Pull request by karly-quick**
snyk-top-banner

#### Snyk changed the following file(s):

- `tt-train/tools/weight-utils/requirements.txt`

#### Breaking Change Risk

Merge Risk: High

> **Notice:** This assessment is enhanced by AI.

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/tt-test-rust/project/4aa8cc94-72b9-4901-bdae-903e8d6a0525?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/tt-test-rust/project/4aa8cc94-72b9-4901-bdae-903e8d6a0525?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Deserialization of Untrusted Data](https://learn.snyk.io/lesson/insecure-deserialization/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":"high","FF_showPullRequestBreakingChanges":true,"FF_showPullRequestBreakingChangesWebSearch":false,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"transformers","from":"4.51.3","to":"5.3.0"}],"env":"prod","issuesToFix":["SNYK-PYTHON-TRANSFORMERS-16876613"],"prId":"20af19b1-d621-4c1e-b29a-cecd4ea32994","prPublicId":"20af19b1-d621-4c1e-b29a-cecd4ea32994","packageManager":"pip","priorityScoreList":[148],"projectPublicId":"4aa8cc94-72b9-4901-bdae-903e8d6a0525","projectUrl":"https://app.snyk.io/org/tt-test-rust/project/4aa8cc94-72b9-4901-bdae-903e8d6a0525?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-TRANSFORMERS-16876613"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by karly-quick** [comment]
Merge Risk: High

The upgrade from `transformers` v4 to v5 is a major version change with significant breaking changes. Projects using TensorFlow or JAX/Flax will break, as v5 is now PyTorch-only. Other critical changes affect model loading, quantization, and authentication.  
  
**Key Breaking Changes:**  
  
*   **PyTorch-Only Backend:** Support for TensorFlow (`TFAutoModel`, `TFBertModel`, etc.) and JAX/Flax (`FlaxAutoModel`) has been completely removed. All models and code must be migrated to PyTorch.  
*   **Quantization API:** The `load_in_4bit=True` and `load_in_8bit=True` parameters in `from_pretrained` are removed. Users must now use the `quantization_config` parameter with a `BitsAndBytesConfig` object.  
*   **Authentication Parameter:** The `use_auth_token` argument is deprecated and replaced by `token` in methods like `from_pretrained`.  
*   **Python Version:** `transformers` v5 requires Python 3.9 or higher.  
*   **Removed Features:** Several less-used features have been removed, including head masking, head pruning, and relative positional biases in BERT-like models.  
*   **Image Processing:** `AutoFeatureExtractor` classes are deprecated in favor of `ImageProcessor`.  
  
**Recommendation:**  
This is a high-risk upgrade that requires significant code refactoring, especially for non-PyTorch users. A direct upgrade is not possible for TensorFlow or JAX/Flax projects. It is critical to review the official migration guide and adapt your code to the new PyTorch-only APIs, updated quantization methods, and other breaking changes before merging.  
  
Source: [Migration Guide](https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQF1WLN5MKV_gCZ8Ms6mvGGHsr70L-A-HnNKZnHNpFwCIZpW85lpNlvEaUzLg_qP02PGXdfYf2LrpFn4SAwletSjjyuIEE_calFVnMw0WivpAGwPU65Yi54dZxbfTRQ7UXXG3Xv0iKwq-n7ER78CXv419IKegU-x4Gk7V2Kk9jtVZaP53GSRIAQC5u_dqLgazUOK3YNxtgzPNq4U0z5FPg==), [Release Notes](https://github.com/huggingface/transformers/releases)

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.

**Comment by karly-quick** [comment]
### :warning: **Snyk checks are incomplete.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :warning: | Open Source Security| 0 | 0 | 0 | 0 | [ See details ](https://app.snyk.io/org/tt-test-rust/pr-checks/be8352b2-124d-47bb-9a15-82a3cbb99449?refs=68fb431f6d0892362765bc15c30503a6986206ad&source=prChecksComment) |
| :warning: | Licenses| 0 | 0 | 0 | 0 | [ See details ](https://app.snyk.io/org/tt-test-rust/pr-checks/be8352b2-124d-47bb-9a15-82a3cbb99449/license?refs=68fb431f6d0892362765bc15c30503a6986206ad&source=prChecksComment) |
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/project/c45c259d-ddf0-4be1-960e-6f383f9dbef2/pr-check/0726b9e3-b9e4-4bcb-800e-7d2489bc8a4a?refs=68fb431f6d0892362765bc15c30503a6986206ad&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).
